mobius/src/orthogonal_arcs.rs

```diff
@@ -0,0 +1,176 @@
+use std::f64::consts::{PI, TAU};
+
+use crate::{
+    geometry::{ArcAngles, Circle, CircularArc, GeneralizedCircle, Line, LineSegment},
+    nearly::is_nearly,
+    Complex,
+};
+
+#[derive(Debug)]
+pub enum OrthogonalArcError {
+    DuplicatePoint(Complex),
+    PointAtCenter(Complex),
+}
+
+pub fn compute_orthogonal_circle(
+    circle: Circle,
+    a: Complex,
+    b: Complex,
+) -> Result<GeneralizedCircle, OrthogonalArcError> {
+    if a == b {
+        return Err(OrthogonalArcError::DuplicatePoint(a));
+    }
+
+    if a == circle.center {
+        return Err(OrthogonalArcError::PointAtCenter(a));
+    }
+
+    if b == circle.center {
+        return Err(OrthogonalArcError::PointAtCenter(b));
+    }
+
+    // If the arc was exactly half a circle, the result will be a line, not a
+    // circle, so let's check for that first.
+    let angle_a = circle.get_angle(a).unwrap();
+    let angle_b = circle.get_angle(b).unwrap();
+
+    // Since we're comparing with half a circle, it doesn't matter if we
+    // measure the difference clockwise or counterclockwise.
+    if is_nearly((angle_a - angle_b).rem_euclid(TAU), PI) {
+        return Ok(GeneralizedCircle::Line(Line::from_points(b, a).unwrap()));
+    }
+
+    // In the past, I computed this for a unit circle
+    // https://github.com/ptrgags/p5-sketchbook/tree/main/HyperbolicConnections#method-2-kite-analysis
+    // but here I need the formula where the first circle is a circle with
+    // radius r1.
+    //
+    // Following the steps of the derivation with the new radius and
+    // center gives the following updated formulas
+    //
+    // r = pq/(2r1)
+    // p = sqrt((4r^4)/(4r^2-q^2))
+    //
+    // finally, the new circle center now is offset by the first
+    // circle's center.
+    //
+    // TODO: Write a better explainer for this.
+
+    let q = (a - b).mag();
+    let r1 = circle.radius;
+    let double_r1 = 2.0 * r1;
+    let denominator = (double_r1 - q) * (double_r1 + q);
+    let p = double_r1 * r1 * (1.0 / denominator).sqrt();
+    let orthog_radius = 0.5 * p * q / r1;
+
+    let angle_bisector = 0.5 * (angle_a + angle_b);
+    let angle_bisector = if (angle_b - angle_a) % TAU > PI {
+        (angle_bisector + PI) % TAU
+    } else {
+        angle_bisector
+    };
+
+    let orthog_center = circle.center + Complex::from_polar(p, angle_bisector);
+    let orthog_circle = Circle {
+        center: orthog_center,
+        radius: orthog_radius,
+    };
+
+    Ok(GeneralizedCircle::Circle(orthog_circle))
+}
+
+pub fn compute_orthogonal_arc(arc: CircularArc, a: Complex, b: Complex) -> CircularArc {
```

Comment:
Similarly here. The arc already specifies angles, get the points from there.

If crinkle_arc needs to adjust the angles to call this method, I think that's okay.

Reply:
This worked out nicely!